src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HealthHeartbeatProvider.cs

```diff
@@ -0,0 +1,226 @@
+        /// The payload items to send out with each health heartbeat.
+        /// </summary>
+        private Dictionary<string, IHealthHeartbeatPayloadExtension> payloadItems;
+        
+        private bool disposedValue = false; // To detect redundant calls to dispose
+        private int intervalBetweenHeartbeatsMs; // time between heartbeats emitted specified in milliseconds
+        private string enabledHeartbeatPayloadFields; // string containing fields that are enabled in the payload. * means everything available.
+        private TelemetryClient telemetryClient; // client to use in sending our heartbeat
+        private int heartbeatsSent;
+
+        public HealthHeartbeatProvider() : this(DefaultHeartbeatIntervalMs, DefaultAllowedFieldsInHeartbeatPayload)
+        {
+        }
+
+        public HealthHeartbeatProvider(int delayMs) : this(delayMs, DefaultAllowedFieldsInHeartbeatPayload)
+        {
+        }
+
+        public HealthHeartbeatProvider(string allowedPayloadFields) : this(DefaultHeartbeatIntervalMs, allowedPayloadFields)
+        {
+        }
+
+        public HealthHeartbeatProvider(int delayMs, string allowedPayloadFields)
+        {
+            this.enabledHeartbeatPayloadFields = allowedPayloadFields;
+            this.intervalBetweenHeartbeatsMs = delayMs;
+            this.payloadItems = new Dictionary<string, IHealthHeartbeatPayloadExtension>();
+            this.heartbeatsSent = 0; // count up from construction time
+        }
+
+        public int HeartbeatIntervalMs => this.intervalBetweenHeartbeatsMs;
+
+        public string EnabledPayloadFields => this.enabledHeartbeatPayloadFields;
+
+        public bool Initialize(TelemetryConfiguration configuration, int? delayMs = null, string allowedPayloadFields = null)
+        {
+            if (this.telemetryClient == null)
+            {
+                this.telemetryClient = new TelemetryClient(configuration);
+            }
+
+            this.intervalBetweenHeartbeatsMs = delayMs.GetValueOrDefault(this.intervalBetweenHeartbeatsMs);
+
+            if (!string.IsNullOrEmpty(allowedPayloadFields))
+            {
+                this.enabledHeartbeatPayloadFields = allowedPayloadFields;
+            }
+
+            this.AddDefaultPayloadItems(this.payloadItems);
+
+            return true;
+        }
+
+        public void RegisterHeartbeatPayload(IHealthHeartbeatPayloadExtension payloadProvider)
+        {
+            if (payloadProvider == null)
+            {
+                throw new ArgumentNullException(nameof(payloadProvider));
+            }
+
+            if (string.IsNullOrEmpty(payloadProvider.Name) || this.payloadItems.ContainsKey(payloadProvider.Name))
+            {
+                throw new ArgumentNullException(nameof(payloadProvider), "Name member of IHealthHeartbeatPayloadExtension must be set to a unique value and cannot be empty");
+            }
+
+            this.payloadItems.Add(payloadProvider.Name, payloadProvider);
```

Comment:
accepting and storing interfaces is dangerous. Make sure every call to this interface is vetted and will not even remotely lead to exception or other negative effects. 

As I mentioned earlier consider exposing specific types to allow to set state an add properties. This is better from defensive programming standpoint

Reply:
Yes, design tweaking is in order here. TBD.